core/tests/recovery-test/tests/genesis-recovery.test.ts

```diff
@@ -0,0 +1,271 @@
+    console.log('Using external node env profile', externalNodeEnvProfile);
+    let externalNodeEnv: { [key: string]: string } = {
+        ...process.env,
+        ZKSYNC_ENV: externalNodeEnvProfile,
+        EN_SNAPSHOTS_RECOVERY_ENABLED: 'false'
+    };
+
+    let mainNode: zkweb3.Provider;
+    let externalNode: zkweb3.Provider;
+    let externalNodeLogs: FileHandle;
+    let externalNodeProcess: ChildProcess;
+    let externalNodeBatchNumber: number;
+
+    before('prepare environment', async () => {
+        expect(process.env.ZKSYNC_ENV, '`ZKSYNC_ENV` should not be set to allow running both server and EN components')
+            .to.be.undefined;
+        mainNode = new zkweb3.Provider('http://127.0.0.1:3050');
+        externalNode = new zkweb3.Provider('http://127.0.0.1:3060');
+        await killExternalNode();
+    });
+
+    let fundedWallet: zkweb3.Wallet;
+
+    before('create test wallet', async () => {
+        const testConfigPath = path.join(process.env.ZKSYNC_HOME!, `etc/test_config/constant/eth.json`);
+        const ethTestConfig = JSON.parse(await fs.readFile(testConfigPath, { encoding: 'utf-8' }));
+        const mnemonic = ethTestConfig.test_mnemonic as string;
+        fundedWallet = zkweb3.Wallet.fromMnemonic(mnemonic, "m/44'/60'/0'/0/0").connect(mainNode);
+    });
+
+    after(async () => {
+        if (externalNodeProcess) {
+            externalNodeProcess.kill();
+            await killExternalNode();
+            await externalNodeLogs.close();
+        }
+    });
+
+    step('generate new batches if necessary', async () => {
+        let pastL1BatchNumber = await mainNode.getL1BatchNumber();
+        while (pastL1BatchNumber < CATCH_UP_BATCH_COUNT) {
+            const transactionResponse = await fundedWallet.transfer({
+                to: fundedWallet.address,
+                amount: 1,
+                token: zksync.utils.ETH_ADDRESS
+            });
+            console.log('Generated a transaction from funded wallet', transactionResponse);
+            const receipt = await transactionResponse.wait();
+            console.log('Got finalized transaction receipt', receipt);
+
+            // Wait until an L1 batch number with the transaction is sealed.
+            let newL1BatchNumber: number;
+            while ((newL1BatchNumber = await mainNode.getL1BatchNumber()) <= pastL1BatchNumber) {
+                await sleep(1000);
+            }
+            console.log(`Sealed L1 batch #${newL1BatchNumber}`);
+            pastL1BatchNumber = newL1BatchNumber;
+        }
+    });
+
+    step('drop external node database', async () => {
+        const childProcess = spawn('zk db reset', {
+            cwd: homeDir,
+            stdio: 'inherit',
+            shell: true,
+            env: externalNodeEnv
+        });
+        try {
+            await waitForProcess(childProcess);
+        } finally {
+            childProcess.kill();
+        }
+    });
+
+    step('drop external node storage', async () => {
```

Comment:
It is, but I feel it will spaghettify the test code even more; as you've written, there's a problem with mixing different levels of abstraction in the test already.